### PR TITLE
optimize circleCI

### PR DESCRIPTION
**Pull request by bunsenstraat**
- add split paralel by previous runtimes. it's based on names in the results json.
- replace pauses as much as possible by assertions ( still a lot to do there ). we should also avoid using assert.containsText for elements in react. but use xpath selectors with contains attribute and then wait on it. this needs more work.
- fix some flaky tests. among others force recompiling when imports are slow.
- use only compilers on localhost by downloading them in the production assets at e2e build. a script manages this and greps the test files for soljson nightly references. all major versions are automatically included. but only for e2e.
- an env variable is passed to the build process to tell the app to use another url for loading compilers.
- use matrix configuration in circle. see below.
- automatic splitting of tests into jobs that have a separate generic job - containing paralel tests - so we can restart easier without having to run **ALL** the tests again. **FASTER!** **SUCCESS!!** if a job only contains a couple of tests it will pass easier.
- runtime is about 7 minutes now
- in config.yml you can find:
```
            parameters:
              browser: ["chrome", "firefox"]
              script: ["browser_test.sh"]
              job: ["0","1","2","3","4","5","6","7","8","9"]
              jobsize: ["10"]
```

- pipe all tests into _**a single result**_: **tests-passed**. this means matrix can be upscaled or downscaled **at will**. when you need to isolate some tests. We will have to config github to just list that test as required.
Screenshot 2022-09-20 at 21 33 59
Screenshot 2022-09-20 at 21 50 56

- each job still has paralel runs but theoretically of course you could have 20 jobs with little or no parallelism. 
Screenshot 2022-09-20 at 21 48 44

- the app created hidden spans containing information about the loading status of the compiler, the editor, the types files etc this is used by e2e to check if the load has completed
- shorten config file with paramater usage

### CODE REVIEW

File: apps/remix-ide-e2e/src/helpers/init.ts [hunk 1]
```diff
@@ -10,38 +10,44 @@ type LoadPlugin = {
 export default function (browser: NightwatchBrowser, callback: VoidFunction, url?: string, preloadPlugins = true, loadPlugin?: LoadPlugin): void {
   browser
     .url(url || 'http://127.0.0.1:8080')
-    .pause(6000)
+    //.pause(6000)
```
Comment (Aniket-Engg) [inline apps/remix-ide-e2e/src/helpers/init.ts:13]: remove this completely

File: apps/remix-ide-e2e/src/helpers/init.ts [hunk 2]
```diff
@@ -10,38 +10,44 @@ type LoadPlugin = {
...
-      .pause(6000)
-      .perform(done())
+      browser
+        .pause(5000)
+        .execute(function (loadPlugin) { // override a plugin url for testing purpose
+          localStorage.setItem('test-plugin-name', loadPlugin.name)
+          localStorage.setItem('test-plugin-url', loadPlugin.url)
+        }, [loadPlugin])
+        .refreshPage()
+        .waitForElementVisible('[data-id="sidePanelSwapitTitle"]')
```
Comment (Aniket-Engg) [inline apps/remix-ide-e2e/src/helpers/init.ts:27]: what is the use of this?

File: apps/remix-ide-e2e/src/tests/ballot.test.ts [hunk 3]
```diff
@@ -131,12 +131,96 @@ module.exports = {
...
       .waitForElementVisible('*[data-id="scFileConfiguration"]', 10000)
       .click('*[data-id="scFileConfiguration"]')
-      .waitForElementVisible('*[data-id="scConfigChangeFilePath"]', 10000)
-      .click('*[data-id="scConfigChangeFilePath"]')
+
+      .waitForElementVisible({
+        selector: '*[data-id="scConfigChangeFilePath"]',
+        abortOnFailure: false
+      }, 10000)
+      .click({
```
Comment (Aniket-Engg) [inline apps/remix-ide-e2e/src/tests/ballot.test.ts:139]: Add a comment explaining the clicking thrice

File: apps/remix-ide-e2e/src/tests/ballot.test.ts [hunk 4]
```diff
@@ -131,12 +131,96 @@ module.exports = {
...
+      .sendKeys('*[data-id="scConfigFilePathInput"]', 'cf.json')
+      .sendKeys('*[data-id="scConfigFilePathInput"]', browser.Keys.ENTER)
+
+      .isVisible({
+        selector:"//*[@class='py-2 remixui_compilerConfigPath' and contains(.,'cf.json')]",
+        suppressNotFoundErrors: true,
+        locateStrategy: 'xpath'
+      }, (okVisible) => {
+        if (!okVisible.value) {
+          browser.waitForElementVisible({
```
Comment (Aniket-Engg) [inline apps/remix-ide-e2e/src/tests/ballot.test.ts:165]: Please add a comment explaining this

File: apps/remix-ide-e2e/src/tests/ballot.test.ts [hunk 5]
```diff
@@ -131,12 +131,96 @@ module.exports = {
...
+          .sendKeys('*[data-id="scConfigFilePathInput"]', 'cf.json')
+          .sendKeys('*[data-id="scConfigFilePathInput"]', browser.Keys.ENTER)
+        }
+      })
+
+      .isVisible({
+        selector:"//*[@class='py-2 remixui_compilerConfigPath' and contains(.,'cf.json')]",
+        suppressNotFoundErrors: true,
+        locateStrategy: 'xpath'
+      }, (okVisible) => {
```
Comment (Aniket-Engg) [inline apps/remix-ide-e2e/src/tests/ballot.test.ts:195]: Please check if it is required actually three times.

File: apps/remix-ide-e2e/src/tests/ballot_0_4_14.test.ts [hunk 6]
```diff
@@ -16,16 +16,21 @@ module.exports = {
   '@sources': function () {
     return sources
   },
-  'Set Ballot 0.4.11': function (browser: NightwatchBrowser) {
+  'Add Ballot #group1 #group2': function (browser: NightwatchBrowser) {
```
Comment (Aniket-Engg) [inline apps/remix-ide-e2e/src/tests/ballot_0_4_14.test.ts:19]: Please check if `#group1 #group2` is required

File: apps/remix-ide-e2e/src/tests/ballot_0_4_14.test.ts [hunk 7]
```diff
@@ -16,16 +16,21 @@ module.exports = {
...
       .clickLaunchIcon('solidity')
-      .setSolidityCompilerVersion('soljson-v0.4.11+commit.68ef5810.js')
+      .setSolidityCompilerVersion('soljson-v0.4.14+commit.c2215d46.js')
       .waitForElementVisible('[for="autoCompile"]')
       .click('[for="autoCompile"]')
       .verify.elementPresent('[data-id="compilerContainerAutoCompile"]:checked')
   },
-  'Compile Ballot with compiler version 0.4.11 #group1 #group2': function (browser: NightwatchBrowser) {
+
+  'Compile Ballot with compiler version 0.4.14 #group1 #group2': function (browser: NightwatchBrowser) {
```
Comment (Aniket-Engg) [inline apps/remix-ide-e2e/src/tests/ballot_0_4_14.test.ts:33]: same here

File: apps/remix-ide-e2e/src/tests/debugger.test.ts [hunk 8]
```diff
@@ -14,8 +14,11 @@ module.exports = {
 
   'Should launch debugger #group1': function (browser: NightwatchBrowser) {
     browser.addFile('blah.sol', sources[0]['blah.sol'])
+      .pause(4000)
+      .clickLaunchIcon('solidity').click('*[data-id="compilerContainerCompileBtn"]')
```
Comment (Aniket-Engg) [inline apps/remix-ide-e2e/src/tests/debugger.test.ts:18]: Please add comment explaining why you need to compile it manually when auto compile is on

File: apps/remix-ide-e2e/src/tests/debugger.test.ts [hunk 9]
```diff
@@ -14,8 +14,11 @@ module.exports = {
...
+      .pause(4000)
       .clickLaunchIcon('udapp')
-      .waitForElementPresent('*[title="Deploy - transact (not payable)"]', 65000)
+      .waitForElementPresent('*[title="Deploy - transact (not payable)"]', 180000)
```
Comment (Aniket-Engg) [inline apps/remix-ide-e2e/src/tests/debugger.test.ts:21]: please review timeout

File: apps/remix-ide-e2e/src/tests/erc721.test.ts [hunk 10]
```diff
@@ -28,6 +28,30 @@ module.exports = {
       .waitForElementVisible('*[data-id="treeViewLitreeViewItemcontracts"]')
       .waitForElementVisible('*[data-id="treeViewLitreeViewItemcontracts/MyToken.sol"]')
       .openFile('contracts/MyToken.sol')
+      .clickLaunchIcon('solidity')
```
Comment (Aniket-Engg) [inline apps/remix-ide-e2e/src/tests/erc721.test.ts:31]: Please add a comment

File: apps/remix-ide-e2e/src/tests/proxy.test.ts [hunk 11]
```diff
@@ -13,12 +14,35 @@ module.exports = {
...
 
-  'Should show deploy proxy option for UUPS upgradeable contract': function (browser: NightwatchBrowser) {
+
+  'Should show deploy proxy option for UUPS upgradeable contract #group1': function (browser: NightwatchBrowser) {
     browser
       .addFile('myTokenV1.sol', sources[0]['myTokenV1.sol'])
       .clickLaunchIcon('solidity')
       .pause(2000)
       .click('[data-id="compilerContainerCompileBtn"]')
+      .clickLaunchIcon('filePanel')
```
Comment (Aniket-Engg) [inline apps/remix-ide-e2e/src/tests/proxy.test.ts:24]: add a comment

File: apps/remix-ide-e2e/src/tests/solidityUnittests.test.ts [hunk 12]
```diff
@@ -271,6 +271,7 @@ module.exports = {
 
   'Debug tests using debugger #group7': function (browser: NightwatchBrowser) {
     browser
+      //.pause()
```
Comment (Aniket-Engg) [inline apps/remix-ide-e2e/src/tests/solidityUnittests.test.ts:274]: remove

File: apps/remix-ide-e2e/src/tests/transactionExecution.test.ts [hunk 13]
```diff
@@ -23,7 +23,7 @@ module.exports = {
           status: 'true Transaction mined and execution succeed',
           'decoded output': { 0: 'uint256: 8' }
         })
-      .pause(120000)
+      //.pause(5000)
```
Comment (Aniket-Engg) [inline apps/remix-ide-e2e/src/tests/transactionExecution.test.ts:26]: please remove

File: apps/remix-ide-e2e/src/tests/usingWebWorker.test.ts [hunk 14]
```diff
@@ -20,21 +20,21 @@ const sources = [
...
     return sources
   },
-  'Using Web Worker': function (browser: NightwatchBrowser) {
+  'Using Web Worker #group1': function (browser: NightwatchBrowser) {
     browser
-      .waitForElementVisible('*[data-id="remixIdeIconPanel"]', 10000)
-      .clickLaunchIcon('filePanel')
-      .addFile('basic.sol', sources[0]['basic.sol'])
-      .clickLaunchIcon('solidity')
+      .waitForElementVisible('[for="autoCompile"]')
```
Comment (Aniket-Engg) [inline apps/remix-ide-e2e/src/tests/usingWebWorker.test.ts:32]: add an explanation comment for disabling auto compile

File: apps/remix-ide/ci/browser_test.sh [hunk 15]
```diff
@@ -14,9 +16,12 @@ yarn run remixd &
...
 
 yarn run build:e2e
-TESTFILES=$(grep -IRiL "@disabled" "dist/apps/remix-ide-e2e/src/tests" | grep "\.spec\|\.test" | sort | circleci tests split )
+# grep -IRiL "@disabled" "dist/apps/remix-ide-e2e/src/tests" | grep "\.spec\|\.test" | xargs -I {} basename {} .test.js | grep -E "\b[${2}]"
+# TESTFILES=$(grep -IRiL "@disabled" "dist/apps/remix-ide-e2e/src/tests" | grep "\.spec\|\.test" | xargs -I {} basename {} .test.js | grep -E "\b[$2]" | circleci tests split --split-by=timings )
+node apps/remix-ide/ci/splice_tests.js $2 $3
+TESTFILES=$(node apps/remix-ide/ci/splice_tests.js $2 $3 | circleci tests split --split-by=timings)
 for TESTFILE in $TESTFILES; do
-    npx nightwatch --config dist/apps/remix-ide-e2e/nightwatch.js $TESTFILE --env=$1  || TEST_EXITCODE=1
+    npx nightwatch --config dist/apps/remix-ide-e2e/nightwatch.js dist/apps/remix-ide-e2e/src/tests/${TESTFILE}.js --env=$1  || TEST_EXITCODE=1
```
Comment (Aniket-Engg) [inline apps/remix-ide/ci/browser_test.sh:24]: please review if this change is required

File: apps/remix-ide/ci/download_e2e_assets.js [hunk 16]
```diff
@@ -0,0 +1,51 @@
+const testFolder = './apps/remix-ide-e2e/src/tests/';
+const fs = require('fs');
+
+let url = 'https://binaries.soliditylang.org/wasm/list.json'
+let request = require('request');
```
Comment (Aniket-Engg) [inline apps/remix-ide/ci/download_e2e_assets.js:5]: `request` has been deprecated

File: libs/remix-ui/editor/src/lib/remix-ui-editor.tsx [hunk 17]
```diff
@@ -606,6 +606,9 @@ export const EditorUI = (props: EditorUIProps) => {
       }
       return result
     }
+    const loadedElement = document.createElement('span')
```
Comment (Aniket-Engg) [inline libs/remix-ui/editor/src/lib/remix-ui-editor.tsx:609]: please add `// for e2e tests` here too

File: libs/remix-ui/solidity-compiler/src/lib/compiler-container.tsx [hunk 18]
```diff
@@ -843,26 +854,27 @@ export const CompilerContainer = (props: CompilerContainerProps) => {
                 }
               }}
             />
-            { !showFilePathInput && <button disabled={!state.useFileConfiguration} data-id="scConfigChangeFilePath" className="btn-secondary" onClick={() => {setShowFilePathInput(true)}}>Change</button> }
+            {!showFilePathInput && <button disabled={!state.useFileConfiguration} data-id="scConfigChangeFilePath" className="btn-secondary" onClick={() => { setShowFilePathInput(true) }}>Change</button>}
           </div>
         </div>
         <div className="px-4">
+          {compilerLoadedVersion ? <span data-id='compilerLoadedVersion' data-version={compilerLoadedVersion}></span> : ''}
```
Comment (Aniket-Engg) [inline libs/remix-ui/solidity-compiler/src/lib/compiler-container.tsx:861]: please check if it is required

File: package.json [hunk 19]
```diff
@@ -66,7 +67,7 @@
     "nightwatch_local_firefox": "yarn run build:e2e && nightwatch --config dist/apps/remix-ide-e2e/nightwatch.js --env=firefox",
     "nightwatch_local_chrome": "yarn run build:e2e && nightwatch --config dist/apps/remix-ide-e2e/nightwatch.js --env=chrome",
     "nightwatch_local_ballot": "yarn run build:e2e && nightwatch --config dist/apps/remix-ide-e2e/nightwatch.js dist/apps/remix-ide-e2e/src/tests/ballot.test.js --env=chrome",
-    "nightwatch_local_ballot_0_4_11": "yarn run build:e2e && nightwatch --config dist/apps/remix-ide-e2e/nightwatch.js dist/apps/remix-ide-e2e/src/tests/ballot_0_4_11.test.js --env=chrome",
+    "nightwatch_local_ballot_0_4_11": "yarn run build:e2e && nightwatch --config dist/apps/remix-ide-e2e/nightwatch.js dist/apps/remix-ide-e2e/src/tests/ballot_0_4_14.test.js --env=chrome",
```
Comment (Aniket-Engg) [inline package.json:70]: please update the name too with `0_4_14`

File: apps/remix-ide-e2e/src/commands/goToVMTraceStep.ts [hunk 20]
```diff
@@ -17,7 +17,10 @@ function goToVMtraceStep (browser: NightwatchBrowser, step: number, incr: number
     .execute((step) => {
       (document.querySelector('*[data-id="slider"]') as any).internal_onmouseup({ target: { value: step }})
     }, [step])
-    .pause(10000)
+    .waitForElementVisible({
+      locateStrategy: 'xpath',
+      selector: `//*[@data-id="treeViewLivm trace step" and contains(.,"${step}")]`,
```
[the remark this answers is not in the document]
Comment (bunsenstraat) [reply apps/remix-ide-e2e/src/commands/goToVMTraceStep.ts:22]: yes

File: apps/remix-ide-e2e/src/commands/goToVMTraceStep.ts [hunk 21]
```diff
@@ -17,7 +17,10 @@ function goToVMtraceStep (browser: NightwatchBrowser, step: number, incr: number
     .execute((step) => {
       (document.querySelector('*[data-id="slider"]') as any).internal_onmouseup({ target: { value: step }})
     }, [step])
-    .pause(10000)
+    .waitForElementVisible({
```
[the remark this answers is not in the document]
Comment (bunsenstraat) [reply apps/remix-ide-e2e/src/commands/goToVMTraceStep.ts:20]: this is just extension of the isVisible action. would not make a difference.

File: apps/remix-ide-e2e/src/commands/addFile.ts [hunk 22]
```diff
@@ -42,10 +46,13 @@ function addFile (browser: NightwatchBrowser, name: string, content: NightwatchC
           .waitForElementContainsText('*[data-id$="/blank"]', '', 60000)
           .sendKeys('*[data-id$="/blank"] .remixui_items', name)
           .sendKeys('*[data-id$="/blank"] .remixui_items', browser.Keys.ENTER)
-          .pause(2000)
-          .waitForElementVisible(`li[data-id="treeViewLitreeViewItem${name}"]`, 60000)
+          .isVisible({
```
[the remark this answers is not in the document]
Comment (bunsenstraat) [reply apps/remix-ide-e2e/src/commands/addFile.ts:51]: I think what the problem is that waitForElementVisible first determines the ID of the element and then polls it with isDplayed https://www.selenium.dev/selenium/docs/api/java/org/openqa/selenium/WebElement.html#isDisplayed--
with an interval, and it seems that if React refreshes the page the nightwatch polling command actually retries to calls the protocol action with the wrong ID resulting in an error. isVisible directly uses the isDisplayed command without polling, it's just a direct webdriver command

File: apps/remix-ide/ci/browser_test.sh (hunk 15 above)
Comment (bunsenstraat) [reply apps/remix-ide/ci/browser_test.sh:24]: yes it's required

File: apps/remix-ide/ci/download_e2e_assets.js (hunk 16 above)
Comment (bunsenstraat) [reply apps/remix-ide/ci/download_e2e_assets.js]: now using axios

**Review by Aniket-Engg — APPROVED** [review]

**Pull request merged**